flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/RestartingTest.java

```diff
@@ -0,0 +1,152 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.declarative;
+
+import org.apache.flink.api.common.JobStatus;
+import org.apache.flink.runtime.JobException;
+import org.apache.flink.runtime.client.JobExecutionException;
+import org.apache.flink.runtime.executiongraph.ExecutionGraph;
+import org.apache.flink.runtime.executiongraph.TestingExecutionGraphBuilder;
+import org.apache.flink.runtime.scheduler.ExecutionGraphHandler;
+import org.apache.flink.runtime.scheduler.OperatorCoordinatorHandler;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+
+import java.time.Duration;
+import java.util.function.Consumer;
+
+import static org.apache.flink.runtime.scheduler.declarative.WaitingForResourcesTest.assertNonNull;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+/** Tests for the {@link Restarting} state of the declarative scheduler. */
+public class RestartingTest extends TestLogger {
+
+    @Test
+    public void testOnEnter() throws Exception {
+        try (MockRestartingContext ctx = new MockRestartingContext()) {
+            Restarting restarting = getRestartingState(ctx);
+            ctx.setExpectedStateChecker((state) -> state == restarting);
+            ctx.setExpectWaitingForResources();
+            restarting.onEnter();
+        }
+    }
+
+    @Test
+    public void testCancel() throws Exception {
+        try (MockRestartingContext ctx = new MockRestartingContext()) {
+            Restarting restarting = getRestartingState(ctx);
+            ctx.setExpectCancelling(assertNonNull());
+            restarting.cancel();
+        }
+    }
+
+    @Test
+    public void testSuspend() throws Exception {
+        try (MockRestartingContext ctx = new MockRestartingContext()) {
+            Restarting restarting = getRestartingState(ctx);
+            ctx.setExpectedStateChecker((state) -> state == restarting);
+            ctx.setExpectFinished(
+                    archivedExecutionGraph ->
+                            assertThat(archivedExecutionGraph.getState(), is(JobStatus.SUSPENDED)));
+            restarting.suspend(new RuntimeException("suspend"));
+        }
+    }
+
+    @Test
+    public void testGlobalFailure() throws Exception {
+        try (MockRestartingContext ctx = new MockRestartingContext()) {
+            Restarting restarting = getRestartingState(ctx);
+            restarting.handleGlobalFailure(new RuntimeException());
+            // no state transition expected
```

Comment:
Can we make this explicit by `ctx.checkNotStateTransition()` or so? Internally we could check the `StateValidators` that nothing has happened.